wordpress-mobile/WordPress-iOS
# [Modal Layout Picker] Populate UI for WPCOM sites with data from the API

**Fixes** https://github.com/wordpress-mobile/gutenberg-mobile/issues/2443
**Related PRs:**
- `gutenberg` https://github.com/WordPress/gutenberg/pull/24702
- `gutenberg-mobile` https://github.com/wordpress-mobile/gutenberg-mobile/pull/2564

## To test:
<details>
<summary><strong>To disable or enable the development version of Modal Layout Picker</strong></summary>

- Open the app from the build that allows FeatureFlags such as a PR build or a local development build
    - By default, the modal layout picker will be disabled.
    - From the site page:
        - Click on your Gravatar.
        - Click on App Settings.
        - Click on Debug.
        - Toggle "Gutenberg Modal Layout Picker" to enable or disable the picker
            <img width=300 src="https://user-images.githubusercontent.com/3384451/90816133-c8b10580-e2f9-11ea-8584-6bbfd7e523e2.gif" />
</details>

When enabled, the Layout Picker should show when creating a new page from My Site and from the Page list.

This PR is for WPCOM sites only self-hosted sites will be supported as part of https://github.com/wordpress-mobile/gutenberg-mobile/issues/2455

### Regular Flow
1. With a good network connection
1. Return to the app and select a WPCOM site
1. Open the Modal Layout Picker
1. **Expect** to see a brief ghost of while the data is fetched
1. **Expect** to see the screen updated with the fetched layouts
    - Layouts in the filter bar should be in alphabetical order
    - The sections in the table view should match the order of the categories in the filter bar
1. You should be able to select a layout from a category. 

### Slow Network
1. Using the [network conditioner](https://mattmiklic.com/2018/02/19/enabling-network-link-conditioner-on-iphone-and-ipad/) set the profile to some less-than-optimal option.
1. Return to the app and select a WPCOM site
1. Open the Modal Layout Picker.
1. **Expect** to see a ghost of empty cells while the data is fetched.
    - You should not be able to select a cell or category while the data is being fetched.
1. **Expect** to see the screen updated with the fetched layouts
    - Layouts in the filter bar should be in alphabetical order
    - The sections in the table view should match the order of the categories in the filter bar
1. You should be able to select a layout from a category. 

### No Network
1. Diable you're device's connection by enabling airplane mode and/or disabling WiFi.
1. Return to the app and select a WPCOM site
1. Open the Modal Layout Picker.
1. **Expect** to see a ghost of empty cells while the data is fetched.
    - You should not be able to select a cell or category while the data is being fetched.
1. You should be able to still create a blank page
    - 📝We will be using cached data for this flow in the future.

## Screenshots:

![FetchLiveData](https://user-images.githubusercontent.com/3384451/90933446-86f18f00-e3cd-11ea-8075-0dff062660a1.gif)


## Additional Context: 

This is a continuation of the work done in https://github.com/wordpress-mobile/WordPress-iOS/pull/14631 and will be followed up by other tickets as part of [Modal Layout Picker Project](https://github.com/wordpress-mobile/gutenberg-mobile/issues?q=is%3Aissue+is%3Aopen+%5BModal+Layout+Picker%5D)

The image sizes will still be optimized as part of https://github.com/wordpress-mobile/gutenberg-mobile/issues/2445 so a less than optimal experience for those shouldn't be a blocker for this step.

If you want to monitor the request via a network inspector you'll be looking for a request like `https://public-api.wordpress.com/wpcom/v2/sites/{{site_id}}/block-layouts`

## PR submission checklist:
